### DISCUSSION TITLE
Any way to get window size and position in labwc?

### DISCUSSION THREAD

**Discussion by stefonarch** [labwc/labwc#1019] [Q&A] [answered]
Context: I [forked](https://github.com/stefonarch/wshot) a little qarma/zenity GUI for screenshots, under sway, wayfire and hyprland "window" caption is doable.

A `labwc -windowinfo` option similar to `wf-info`?

**Comment by Consolatis** [comment, marked as answer]
For X11 windows you may get around using some X11 API. For wayland windows: no chance as long as
- ext-screencopy isn't updated to support the ext-foreign-toplevel-list protocol
  - merged into wayland-protocols
  - implemented into wlroots
  - labwc using that wlroots version
- or: labwc suddenly decides to implement custom IPC (e.g. *really* unlikely)

However, using a tool like slurp (or its internal protocols) the user could select the region they want to capture rather then selecting a window.

**Reply by stefonarch** [reply]
Sure, slurp can be used, but it's hard to hit exactly the borders though.
 I was wondering if OSD would [eventually display window size](https://github.com/labwc/labwc/issues/1013) if this could be accessed too, together with x,y, to feed it to` grim -g `.

**Reply by Consolatis** [reply]
Well, displaying it is different to having an external application read it. Unless you intend to take a screenshot of the OSD, run some OCR software on it to have it as text and then use it to do the actual screenshot :)

I fear the only official way that labwc can support window captures is via an official protocol. (Short of doing a full screenshot, showing it in fullscreen, allowing the user to click on a "window" in that screenshot, doing edge detection to figure out the window region and then cutting that from the screenshot).

**Reply by stefonarch** [reply]
Well, slurp is still better than _that_ :)

**Comment by thelamer** [comment]
Sorry for posting in a necro discussion but just in general this is going to come up again. XFCE seems to be fully integrating labwc as their wayland solution and it is impossible to support at-spi/A11y without window position and geometry. Blind users will be walled out of wayland without this. With window listing, position, and geometry it is possible to adjust coordinates applications report to their relative 0,0 position where the window is rendered, without it the click coordinate information is useless. 

So even if a custom IPC needs to be developed something needs to expose these internal scene integers, kwin does this via dbus calls.

**Reply by Consolatis** [reply]
I am completely unaware of how at-spi / A11y works. I always speculated that its applications which export a tree of labels to a dbus service. In such a scenario the cursor within the window would be enough for an application to figure out the currently hovered-over label in that tree. How do global cursor positions play a role there?

In general I'd really like labwc to be as inclusive as possible.

**Reply by thelamer** [reply]
In general it is a Wayland problem as a whole. There is no unified solution yet that I know of. When getting the info from a specific application in a Wayland desktop everything is relative that that app and starts at 0,0 of that surface for that app.
My comments was that with an interface to determine window positioning at the DE level you can then extrapolate the absolute coordinates. That is the workaround I landed on for labwc and KDE while writing a custom IPC patch to hook into labwc.

I am leveraging this for bot control of desktop applications, but this is one of those things that "just works" for X11, needs workarounds for KDE Wayland leveraging Kwin dubs, and is blocked on labwc as these are private functions. I don't have the good answer as to how to implement this properly just that it is a hole in the stack and I have to assume blind users are stuck on X11 for now.

**Reply by Consolatis** [reply]
> it is impossible to support at-spi/A11y without window position and geometry. Blind users will be walled out of wayland without this.

> I have to \*assume\* blind users are stuck on X11 for now.

Then please state it as such rather than stating it as a fact.

Abusing an accessibility system to get absolute mouse coordinates for some bot control of a compositor is a completely different thing versus the needs of a blind user.

**Reply by thelamer** [reply]
I mean you can easily confirm it for yourself or just do a simple Google search. Apologies I did not realize using an API that defines screen coordinates for items on screen was abuse.

I just discovered this limitation because of what I was developing otherwise I would have never even known it existed. 

https://github.com/splondike/wayland-accessibility-notes/blob/main/talon-requirements.md#07-get-window-position-and-sise

**Reply by johanmalm** [reply]
We are keen to support accessibility as long as the cost (code complexity, etc) is not completely disproportionate. For example, we consider animations to be one of those disproportionate ones.

Welcome proposals, designs, ideas on increasing accessibility.

It's curious to me that the stated KDE protocol method is `set_minimized_geometry`. Any ideas?

[org_kde_plasma_window::set_minimized_geometry](https://wayland.app/protocols/kde-plasma-window-management#org_kde_plasma_window:request:set_minimized_geometry)

**Reply by thelamer** [reply]
You all will likely not agree with me but I believe agent driven control of a desktop computer mixed with the old manual mode provided by Talon/Orca style solutions is the future for people with disabilities. Telling an agent to open a text editor, dictating notes to it, then having it save and organize the files has a drastically lower barrier of entry than listening to a screen reader for a desktop application that was designed around people with sight.
Just as a thought experiment I closed my eyes and tried to interact with applications and web pages using a screen reader, then did it with TTS and an agent leveraging these same APIs to bypass the inaccuracies of using vision models. The most important part for me at least was asking for it to "explain what is on the screen" or "do you see a way to login on this website". With a screen reader you get literal minutes of "SEPERATOR, INPUT FIELD TAG USERNAME" while hitting tab when trying to use a website in a harsh robotic voice, while with an agent you just get it done in seconds. Now this has nothing to do with labwc, I was just generally making these comments as I do think self hosted or low param agents being able to drive a desktop with these apis has value. 

As far as specifics of how this should be implemented you all would know better than me, I just assume when something like XFCE picks up labwc as their default wayland window manager this band aid has to be ripped off at some point, wlroots gets you 90% of the way there and it looks like GNOME/Kwin have landed on dbus calls that can be made for window positioning and screen readers are either working to adapt or already done so by hitting these calls to determine visibility of the windows or their absolute mouse coordinates to drive clicks and describe screen content. It it not really on labwc to implement all of the features that would make a11y natively report desktop states like it currently works in X11 it is just about getting parity and some kind of interface to hook into and determine visibility with positioning. 

For minimize specifically I would think that would more land on wlr-foreign-toplevel-management and xfce4-panel than it would be something you all would need to worry about.

**Reply by elviosak** [reply]
> when something like XFCE picks up labwc

XFCE is working on their own compositor based on the lib from COSMIC, i think they released an alpha recently, LXQt doesn't have a default and probably won't have it. MATE i think is using wayfire, which is probably the best one to try those things since u can write plugins for it.

**Reply by Consolatis** [reply]
In general I agree to the assumption about agent driven workflows. However, it doesn't need global cursor coordinates. That may currently be a indirect way to get an agent to move the cursor to specific items within an application via the virtual pointer protocol and then execute a click. That is a completely backwards and way too complicated for the actual goal. A simpler approach would be to tell the application `focus text field named 'comment'` or `press button named "Submit"`. The cursor doesn't play a role here at all.

Shifting focus to another window is already possible on wlroots with the wlr-foreign-toplevel-management protocol (example: `wlrctl`, various panels). Applications also do know if they currently have keyboard focus so they can expose that information via dbus and a control application (screenreader or some agent) could use that to answer the question of "where am I" (focused application + focused element within that application).
